packages/lite-insight/src/attribution/multiDimension.ts

```diff
@@ -0,0 +1,136 @@
+    }
+
+    const dimValues: string[] = new Array(stack.length);
+    for (let i = 0; i < stack.length; i += 1) {
+      dimValues[i] = item[stack[i]] as string;
+    }
+    const dimValName = dimValues.join('-');
+    if (!Object.prototype.hasOwnProperty.call(resultTemp[dimName].detail, dimValName)) {
+      const dimInfoSub: disassembleInfoSub = {
+        startValue: 0,
+        endValue: 0,
+        diff: 0,
+        contribution: 0,
+      };
+      resultTemp[dimName].detail[dimValName] = dimInfoSub;
+    }
+
+    const tempDimInfoSub: disassembleInfoSub = resultTemp[dimName].detail[dimValName];
+    resultTemp[dimName].total -= tempDimInfoSub.diff;
+    tempDimInfoSub.startValue += item[baseDimension] === basePointLeft ? (item[measure] as number) : 0;
+    tempDimInfoSub.endValue += item[baseDimension] === basePointRight ? (item[measure] as number) : 0;
+
+    if (!(tempDimInfoSub.startValue === 0 && tempDimInfoSub.endValue === 0)) {
+      tempDimInfoSub.diff = tempDimInfoSub.endValue - tempDimInfoSub.startValue;
+    }
+    resultTemp[dimName].total += tempDimInfoSub.diff;
+    return resultTemp;
+  }
+
+  /** Case1: dimensions[index] is included */
+  stack.push(dimensions[index]);
+  const resultFirst = enumerateAllDimensionCombinationsByDFS(
+    item,
+    index + 1,
+    dimensions,
+    result,
+    stack,
+    measure,
+    baseDimension,
+    basePointLeft,
+    basePointRight
+  );
+  stack.pop();
+
+  /** Case2: dimensions[index] is not included */
+  const resultSecond = enumerateAllDimensionCombinationsByDFS(
+    item,
+    index + 1,
+    dimensions,
+    resultFirst,
+    stack,
+    measure,
+    baseDimension,
+    basePointLeft,
+    basePointRight
+  );
+
+  return resultSecond;
+};
+
+export const attributeSingleMeasure2MultiDimension = (
+  sourceData: Datum[],
+  dimensions: string[],
+  measure: string,
+  baseDimension: string,
+  basePointLeft: string | number,
+  basePointRight: string | number
+) => {
+  /** remove invalid data */
+  const data = sourceData.filter((item) => !Object.values(item).some((v) => v === null || v === undefined));
+
+  let result = {};
+  /** traverse the input data and build the result data structure; */
+  data.forEach((item) => {
+    if (item[baseDimension] === basePointLeft || item[baseDimension] === basePointRight) {
+      const stack: string[] = [];
+      result = enumerateAllDimensionCombinationsByDFS(
+        item,
+        0,
+        dimensions,
+        result,
+        stack,
+        measure,
+        baseDimension,
+        basePointLeft,
+        basePointRight
+      );
+    }
+  });
+
+  /** calculate the contribution of every dimension value combination to the designated measure */
+  Object.values(result).forEach((value: disassembleInfo) => {
+    Object.values(value.detail).map((valueSub) => {
+      const item: disassembleInfoSub = JSON.parse(JSON.stringify(valueSub));
+      item.contribution = valueSub.diff / value.total;
```

Comment:
prefer using the spread syntax ( { ...obj,  newKey: xxx } ) to update an object ,  it doesn't mutate the original object.

Reply:
got it.